### PR TITLE
Add puzzler for overloaded function

### PR DESCRIPTION
**Pull request by ConstOrVar**
@angryziber , we discuss that puzzler at recent mobius conf

### CODE REVIEW

File: src/functions/OverloadedOoops.kts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package functions.overloadedOoops
+
+class Test {
```
Comment (TWiStErRob) [inline src/functions/OverloadedOoops.kts:3]: this class is unnecessary, top level functions will repro the same; it'll make the puzzle shorter and simpler to understand

File: src/functions/OverloadedOoops.kts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    fun test(block: () -> Int = { 0 }): String {
+        return "${block()}"
```
Comment (TWiStErRob) [inline src/functions/OverloadedOoops.kts:6]: thoughts on `block().toString()`?

File: src/functions/OverloadedOoops.kts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+val t = Test()
+val a = t.test { 2 } == t.test2 { 2 }()
+val b = t.test() == t.test2()()
+
+
+{ println("Finish $a $b") }()
+
+// What will it print?
```
Comment (TWiStErRob) [inline src/functions/OverloadedOoops.kts:25]: Would you mind explaining the answer? Did you find this IRL?

File: src/functions/OverloadedOoops.kts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        return "${block()}"
+    }
+
+    fun test2(block: () -> Int = { 0 }): () -> String {
+        return { test(block) }
+    }
+}
+
+
+{ println("Start") }()
```
Comment (TWiStErRob) [inline src/functions/OverloadedOoops.kts:15]: Why IIFE? never seen that in Kotlin before

**Review by TWiStErRob — COMMENTED** [review]
Amazing puzzle, it got me :) Mind, blown!

File: src/functions/OverloadedOoops.kts (hunk 1 above)
Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:3]: @TWiStErRob , yes, you are right. I will remove redundant class declaration

File: src/functions/OverloadedOoops.kts (hunk 2 above)
Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:6]: It's also suitable. Don't see any arguments why not to use that approach. Do you think, it's easier to understand?

Comment (TWiStErRob) [reply src/functions/OverloadedOoops.kts:6]: Yeah, I think it's much simpler: it reads as "call the block" and the "convert result to string". The current expression has a lot of symbols: `"${()}"`, which is tricky to parse, even with highlight.

Comment (TWiStErRob) [reply src/functions/OverloadedOoops.kts:6]: I think it's worth simplifying as that's not the crux of the puzzle, I don't think this puzzle needs a red herring, it's complex enough as it is :) Remember that in conference settings people have only a few seconds to read these, sometimes on bad projections.

File: src/functions/OverloadedOoops.kts (hunk 4 above)
Comment (TWiStErRob) [reply src/functions/OverloadedOoops.kts:15]: I mean I know, why now, just curious how you ended up with an IIFE in Kotlin

File: src/functions/OverloadedOoops.kts (hunk 3 above)
Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:25]: No, that code is not used in production. It's used to make puzzler looks like it's solve real case. I mean, if you write only 
`{ println("Finish $a $b") }`
without calling it, it's really strange a bit for what reason it looks so. 

So, that puzzler was found during development of android application. But construction like `{}()` is not from real life.

Comment (TWiStErRob) [reply src/functions/OverloadedOoops.kts:25]: Ah, ok, makes sense.
Btw, by "explaning the answer" I meant `Rationale.md`, it's quite a tricky one.

File: src/functions/OverloadedOoops.kts (hunk 4 above)
Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:15]: @TWiStErRob Actually, I try to make lambda not to be alone here. That's why it looks so

File: src/functions/OverloadedOoops.kts (hunk 2 above)
Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:6]: Ok, I got it. What do you think if we transform puzzler like that
`fun test(block: () -> Int = { 0 }): String { return block().toString() }`

val a = test { 2 }
val b = test()
    
//     print("Test")
{ 1 }
println("Finish $a $b")

And the question is "What's changed if we uncomment string?"
1) Nothing changed ("2 0")
2) Before "2 1", after "Test 2 0"
3) Code will not compile
4) None of the above

Is it looks better now? Or you still think that it's very complicated?

Comment (ConstOrVar) [reply src/functions/OverloadedOoops.kts:6]: @TWiStErRob , don't you mind if I replace previous variant with that one (described above)? I think it looks simpler, less code, but demonstrates exactly the same problem as initial code.

**Comment by angryziber** [comment]
I have changed it to the simpler version and added the Rationale.
https://github.com/angryziber/kotlin-puzzlers/commit/3e8b1774626546e8e0928899ac25a3a4395297c7
@ConstOrVar @TWiStErRob what do you think, is it still the same puzzler or already another one? :-)

**Pull request merged**